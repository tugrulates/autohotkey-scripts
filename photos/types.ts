/**
 * Represents the EXIF (Exchangeable Image File Format) metadata of a photo.
 *
 * This only lists the fields relevant for my photography workflow.
 */
export interface Exif {
  /** The title of the photo. */
  title?: string;
  /** Text describing the contents of the photo. */
  description?: string;
  /** Keywords for findability. */
  keywords?: string[];
  /** The date the photo was taken. */
  date?: string;
  /** The location where the photo was taken. */
  location?: string;
  /** The camera or phone used to take the photo. */
  camera?: string;
  /** Lens properties that were used to take the photo. */
  lens?: string;
  /** The pixel resolution of the photo. */
  size?: string;
}

/**
 * Represents a photo with additional metadata and sizes.
 *
 * @extends Exif with all fields except resolution, which are listed on individial
 * file sizes instead.
 */
export interface Photo extends Omit<Exif, "size"> {
  /** Exchangable id of the photo. */
  slug: string;
  /** Different size variants of this photo. */
  sizes: {
    file: string;
    description?: string;
    size?: string;
    sameExif: boolean;
  }[];
}
