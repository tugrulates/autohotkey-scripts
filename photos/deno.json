{
  "name": "@tugrulates/photos",
  "version": "0.1.1",
  "exports": "./mod.ts",
  "tasks": {
    "compile": "deno compile -A -o=../dist/ mod.ts"
  }
}
