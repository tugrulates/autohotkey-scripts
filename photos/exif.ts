/**
 * Displays information about the jpg files for a photo directory.
 *
 * ### Usage
 *
 * ```sh
 * deno -A photos/exif.ts [photos...] [--copy] [--json]
 * ```
 *
 * ```
 * üñº Title [‚ö†Ô∏è warnings]
 * ```
 */

import $ from "jsr:@david/dax";
import { parseArgs } from "jsr:@std/cli";
import { basename, dirname, join } from "jsr:@std/path";
import { getOptional } from "../common/cli.ts";
import { Exif, Photo } from "./types.ts";

const SOURCE_FILE = "source.jpg";
const EXIFTOOL = ["exiftool", "-q", "-overwrite_original_in_place"];

/**
 * Returns true if the directory contains a source photo.
 *
 * @param dir Directory to check.
 * @returns True if the directory contains a source photo.
 */
async function isPhotoDirectory(dir: string): Promise<boolean> {
  try {
    await Deno.lstat(`${dir}/${SOURCE_FILE}`);
    return true;
  } catch {
    return false;
  }
}

/**
 * Returns a list of all JPG files in the directory.
 *
 * @param photo Directory to check.
 * @returns List of all JPG files in the directory.
 */
async function getFiles(photo: string): Promise<string[]> {
  if (!(await isPhotoDirectory(photo))) return [photo];
  return (await Array.fromAsync(Deno.readDir(photo)))
    .filter((f) => f.isFile)
    .filter((f) => f.name.endsWith(".jpg"))
    .map((f) => join(photo, f.name))
    .toSorted();
}

/**
 * Returns a list of all photo directories under cwd.
 *
 * @returns List of all photo directories under cwd.
 */
async function* allPhotos(): AsyncGenerator<string> {
  for await (const dirEntry of Deno.readDir(".")) {
    if (dirEntry.isDirectory && await isPhotoDirectory(dirEntry.name)) {
      yield dirEntry.name;
    }
  }
}

/**
 * Returns the EXIF data for the file.
 *
 * @param file File to get EXIF data for.
 * @returns EXIF data for the file.
 */
async function getExif(file: string): Promise<Exif> {
  const lines = await $`${EXIFTOOL} ${file}`.lines();
  const data = Object.fromEntries(
    lines.map((line) => line.match(/^([^:]+?)\s*:\s*(.*?)$/)?.slice(1) ?? []),
  );
  const date = new Date(
    data["Date/Time Original"]
      ?.replace(/:/, "-")
      ?.replace(/:/, "-")
      ?.replace(/ /, "T"),
  );
  return {
    title: data["Title (en)"],
    description: data["Image Description"],
    keywords: data["Keywords"]?.split(",").map((s: string) => s.trim()),
    date: date.toJSON(),
    location: data["Location"],
    camera: data["Camera Model Name"],
    lens: data["Lens Model"],
    size: data["Image Size"],
  };
}

/**
 * Returns the data for a photo or a single file.
 *
 * @param photo Photo or file to get data for.
 * @returns Data for the photo or file.
 */
async function getData(photo: string): Promise<Photo> {
  const files = await getFiles(photo);
  const exif = Object.fromEntries(
    await Promise.all(
      files.map(async (f) => [basename(f), await getExif(f)] as const),
    ),
  );
  const base = exif[SOURCE_FILE] ?? Object.values(exif)[0];
  return {
    slug: basename(photo),
    ...{ ...base, size: undefined },
    sizes: Object.entries(exif).map(([file, data]) => ({
      file,
      description: data.description,
      size: data.size,
      sameExif:
        JSON.stringify({ ...data, description: undefined, size: undefined }) ===
          JSON.stringify({ ...base, description: undefined, size: undefined }),
    })),
  };
}

/**
 * Returns a warning message if the photo is missing required fields.
 *
 * @param data Photo data to check.
 * @returns Message with the following text: [‚ö†Ô∏è warnings...].
 */
function check(data: Photo) {
  const result = [];
  for (
    const field of [
      "title",
      "description",
      "keywords",
      "date",
      "location",
      "camera",
      "lens",
    ] as const
  ) {
    if (!data[field]) result.push(field);
  }
  if (!data.title) result.push("title");
  if (!data.description) result.push("description");
  if (!data.keywords) result.push("keywords");
  if (!data.date) result.push("date");
  if (!data.location) result.push("location");
  if (!data.camera) result.push("camera");
  if (!data.lens) result.push("lens");
  if (!data.sizes.every((size) => size.sameExif)) result.push("same");
  if (result.length) return `[‚ö†Ô∏è ${result.join(", ")}]`;
  return "";
}

/**
 * Copies the EXIF data from source.jpg to all other jpg files.
 *
 * @param photo Directory containing jpg files or a single JPG file.
 */
async function copyExif(photo: string) {
  const base = join(
    await isPhotoDirectory(photo) ? photo : dirname(photo),
    SOURCE_FILE,
  );
  const files = (await getFiles(photo)).filter((f) => f !== base);
  await Promise.all(files.map(async (file) => {
    await $`${EXIFTOOL} -tagsfromfile ${base} -codedcharacterset=UTF8 -all ${file}`;
    let description = await $`${EXIFTOOL} -ImageDescription ${base} -args`
      .text();
    if (description === "") description = "-ImageDescription:";
    const groupMatch = /.*-(\d+).jpg/.exec(file);
    if (groupMatch !== null) description += ` (image ${groupMatch[1]})`;
    await $`${EXIFTOOL} -tagsfromfile ${base} -codedcharacterset=UTF8 ${description} -all ${file}`;
  }));
}

if (import.meta.main) {
  const spec = { boolean: "json" } as const;
  const args = parseArgs(Deno.args, spec);
  const photos = args._.length ? getOptional(args) : allPhotos();

  for await (const photo of photos) {
    if (args.copy) await copyExif(photo);
    const data = await getData(photo);
    if (args.json) console.log(JSON.stringify(data));
    else console.log(`üñº  ${data.title} ${check(data)}`);
  }

  Deno.exit(0);
}
