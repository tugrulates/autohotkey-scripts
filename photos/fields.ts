/** All extracted EXIF fields. */
export const FIELDS = [
  "src",
  "width",
  "height",
  "title",
  "description",
  "keywords",
  "date",
  "location",
  "city",
  "state",
  "country",
  "camera",
  "lens",
  "editing",
  "license",
] as const;

/** EXIF fields that are supposed to be different in variants. */
export const VARIANT_FIELDS = [
  "src",
  "width",
  "height",
  "description",
] as const;

/** EXIF fields that can be omitted. */
export const OPTIONAL_FIELDS = [
  "city",
] as const;
