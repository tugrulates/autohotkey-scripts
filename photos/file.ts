import { join } from "@std/path";

const SOURCE_FILE = "source.jpg";

/**
 * Returns true if the directory contains a source photo.
 *
 * @param dir Directory to check.
 * @returns True if the directory contains a source photo.
 */
export async function isPhotoDirectory(dir: string): Promise<boolean> {
  try {
    await Deno.lstat(`${dir}/${SOURCE_FILE}`);
    return true;
  } catch {
    return false;
  }
}

/**
 * Returns a list of all JPG files in the directory.
 *
 * @param photo Directory to check.
 * @returns List of all JPG files in the directory.
 */
export async function getFiles(photo: string): Promise<string[]> {
  if (!(await isPhotoDirectory(photo))) return [photo];
  return (await Array.fromAsync(Deno.readDir(photo)))
    .filter((f) => f.isFile)
    .filter((f) => f.name.endsWith(".jpg"))
    .map((f) => join(photo, f.name))
    .toSorted();
}

/**
 * Returns a list of all photo directories under cwd.
 *
 * @returns List of all photo directories under cwd.
 */
export async function* allPhotos(): AsyncGenerator<string> {
  for await (const dirEntry of Deno.readDir(".")) {
    if (dirEntry.isDirectory && await isPhotoDirectory(dirEntry.name)) {
      yield dirEntry.name;
    }
  }
}
