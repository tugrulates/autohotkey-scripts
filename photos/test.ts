import { copy } from "@std/fs";
import { basename } from "@std/path";
import { join } from "@std/path/join";
import { afterEach, beforeEach, describe, it } from "@std/testing/bdd";
import { assertSnapshot } from "@std/testing/snapshot";
import { copyExifToVariants, getPhoto, type Photo } from "@tugrulates/photos";

// @todo Fix the timer leaks.
describe("photos", { sanitizeOps: false, sanitizeResources: false }, () => {
  let tmpDirectory: string;

  beforeEach(async () => {
    tmpDirectory = await Deno.makeTempDir();
    await copy("photos/testdata", tmpDirectory, { overwrite: true });
  });

  afterEach(async () => {
    await Deno.remove(tmpDirectory, { recursive: true });
  });

  function trimSource(photo: Photo): Photo {
    return {
      ...photo,
      src: basename(photo.src),
      variants: photo.variants.map((v) => ({
        ...v,
        src: basename(v.src),
      })),
    };
  }

  it("getPhoto()", async (t) => {
    const photo = await getPhoto(
      join(tmpDirectory, "floating-around/source.jpg"),
    );
    await assertSnapshot(t, trimSource(photo));
  });

  it("copyExifToVariants()", async (t) => {
    const before = await getPhoto(
      join(tmpDirectory, "winter-pause/source.jpg"),
    );
    await assertSnapshot(t, trimSource(before));
    await copyExifToVariants(before);
    const after = await getPhoto(join(tmpDirectory, "winter-pause/source.jpg"));
    await assertSnapshot(t, trimSource(after));
  });
});
