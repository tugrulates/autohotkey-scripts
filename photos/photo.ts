import $ from "@david/dax";
import { pick } from "@std/collections";
import { basename, dirname, join } from "@std/path";
import { FIELDS } from "./fields.ts";
import type { Exif, Photo } from "./types.ts";

const EXIFTOOL = ["exiftool", "-q", "-overwrite_original_in_place"];

/**
 * Returns the EXIF data for the file.
 *
 * @param src File to get EXIF data for.
 * @returns EXIF data for the file.
 *
 * @todo This checks if the original date/time has an offset and uses the
 * create date if not. This is a workaround for Affinity Photo not filing
 * timezones.
 */
async function getExif(src: string): Promise<Exif> {
  const lines = await $`${EXIFTOOL} -d "%Y-%m-%dT%H:%M:%S%z" ${src}`.lines();
  const data = Object.fromEntries(
    lines.map((line) => line.match(/^([^:]+?)\s*:\s*(.*?)$/)?.slice(1) ?? []),
  );
  return {
    src,
    width: parseInt(data["Image Width"]),
    height: parseInt(data["Image Height"]),
    title: data["Title (en)"],
    description: data["Image Description"],
    keywords: data["Keywords"]?.split(",").map((s: string) => s.trim()),
    date: data["Offset Time Original"]
      ? data["Date/Time Original"]
      : data["Create Date"],
    location: data["Location"],
    city: data["City"],
    state: data["State"],
    country: data["Country"],
    camera: data["Camera Model Name"],
    lens: data["Lens Model"],
    editing: data["History Software Agent"],
    license: data["License"],
  };
}

/**
 * Copies the EXIF data from photo source to its variants.
 *
 * @param photo Photo data for managing EXIF.
 */
export async function copyExifToVariants(photo: Photo) {
  const baseArgs =
    await $`${EXIFTOOL} -CreateDate -ImageDescription ${photo.src} -args`
      .lines();
  await Promise.all(photo.variants.map(async (variant) => {
    const groupMatch = /.*-(\d+).jpg/.exec(variant.src);
    const args = baseArgs.map((arg) =>
      arg
        .replace(/^-CreateDate=/, "-SubSecDateTimeOriginal=")
        .replace(
          /^(-ImageDescription=.*?)(\.?)$/,
          groupMatch ? `$1 (image ${groupMatch[1]}).` : "$1$2",
        )
    );
    await $`${EXIFTOOL} -tagsfromfile ${photo.src} -codedcharacterset=UTF8 ${args} -all ${variant.src}`;
  }));
}

/**
 * Returns a list of all JPG files in the directory.
 *
 * @param dir Directory to check.
 * @returns List of all JPG files in the directory.
 */
async function getFiles(dir: string): Promise<string[]> {
  return (await Array.fromAsync(Deno.readDir(dir)))
    .filter((f) => f.isFile)
    .filter((f) => f.name.endsWith(".jpg"))
    .map((f) => join(dir, f.name))
    .toSorted();
}

/**
 * Returns the data for a photo using a single file.
 *
 * @param path Source photo to get information for.
 * @returns Data for the photo or file.
 */
export async function getPhoto(path: string): Promise<Photo> {
  const files = await getFiles(dirname(path));
  const exif = await Promise.all(files.map((f) => getExif(f)));
  const photo = exif.find((e) => e.src === path);
  if (!photo) throw new Error(`EXIF cannot be extracted for ${path}.`);
  return {
    id: basename(path),
    ...photo,
    variants: exif.filter((e) => e.src !== path).map((data) => ({
      ...pick(
        data,
        FIELDS.filter((f) =>
          JSON.stringify(photo[f]) !== JSON.stringify(data[f])
        ),
      ),
      src: data.src,
      width: data.width,
      height: data.height,
    })),
  };
}
