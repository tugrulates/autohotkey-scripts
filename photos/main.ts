import { colors } from "@cliffy/ansi/colors";
import { Command } from "@cliffy/command";
import { omit } from "@std/collections/omit";
import { expandGlob } from "@std/fs";
import { join } from "@std/path";
import { OPTIONAL_FIELDS, VARIANT_FIELDS } from "./fields.ts";
import { copyExifToVariants, getPhoto } from "./photo.ts";
import type { Photo } from "./types.ts";

const SOURCE_FILE = "source.jpg";

async function isFile(path: string): Promise<boolean> {
  try {
    const info = await Deno.lstat(path);
    return info.isFile;
  } catch {
    return false;
  }
}

async function* getPhotos(photos: string[]): AsyncGenerator<Photo> {
  if (photos.length === 0) {
    for await (const entry of expandGlob(join("*", SOURCE_FILE))) {
      yield await getPhoto(entry.path);
    }
  }
  for (const photo of photos) {
    if (await isFile(photo)) {
      yield (await getPhoto(photo));
    } else if (await isFile(join(photo, SOURCE_FILE))) {
      yield getPhoto(join(photo, SOURCE_FILE));
    }
  }
}

function check(data: Photo) {
  const result = [];
  for (const [field, value] of Object.entries(omit(data, OPTIONAL_FIELDS))) {
    if (!value) result.push(`missing:${field}`);
  }
  for (const size of data.variants) {
    for (const field in omit(size, VARIANT_FIELDS)) {
      if (field in data) result.push(`${size.src}:${field}`);
    }
  }
  if (result.length) return `[${colors.yellow(result.join(", "))}]`;
  return "";
}

function getCommand() {
  return new Command()
    .name("photos")
    .example("photos", "Lists all photos under current directory.")
    .example("photos [photo] --json", "Data for a photo with all variants.")
    .example("photos [photo] --copy", "Copy EXIF data to all variants.")
    .description("Manage photos.")
    .arguments("[photos...:file]")
    .option("--copy", "Copy the EXIF from source JPG to other variants.")
    .option("--json", "Output the EXIF information as JSON.")
    .action(async ({ copy, json }, ...photos) => {
      for await (let photo of getPhotos(photos)) {
        if (copy) {
          await copyExifToVariants(photo);
          photo = await getPhoto(photo.src);
        }
        if (json) console.log(JSON.stringify(photo));
        else console.log(`ðŸ–¼  ${photo.title} ${check(photo)}`);
      }
    });
}

/** CLI entrypoint. */
export async function main(): Promise<void> {
  const command = getCommand();
  await command.parse();
}
