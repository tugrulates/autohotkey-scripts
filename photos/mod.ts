/**
 * Photography editing and publishing workflow.
 *
 * The binary for `exiftool` is bundled with this library. However, the
 * compiled binary `photos` requires `exiftool` to be present in the system.
 * For macOS, it can be installed using Homebrew with `brew install exiftool`.
 * For other systems, see the [official website](https://exiftool.org/).
 *
 * @module
 */

import { main } from "./main.ts";

export * from "./photo.ts";
export * from "./types.ts";

if (import.meta.main) await main();
