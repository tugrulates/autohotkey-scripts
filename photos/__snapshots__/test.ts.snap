export const snapshot = {};

snapshot[`getPhoto() 1`] = `
{
  camera: "Apple iPhone 13 mini",
  city: "San Francisco",
  country: "United States",
  date: "2023-12-10T14:28:09-08:00",
  description: "Four jellyfish with translucent bodies and long tentacles float gracefully against a vivid background in an aquarium setting.",
  height: 180,
  id: "floating-around",
  keywords: [
    "California Academy Of Sciences",
    "Translucent Bodies",
    "Long Tentacles",
    "Float Gracefully",
    "Vivid Background",
    "Aquarium Setting",
    "Marine Life",
    "Underwater World",
    "Ocean Creatures",
    "Exotic Jellyfish",
    "Aquatic Environment",
    "Sea Life",
    "Ethereal Beauty",
    "Marine Biology",
    "Underwater Exploration",
    "Aquatic Creatures",
    "Jellyfish Swimming",
    "Oceanic Wildlife",
    "Marine Exhibit",
    "Horizontal",
    "No People",
    "Photography",
    "Jellyfish",
    "Underwater",
    "Water",
    "Blue",
    "Undersea",
    "Sea",
    "Nature",
    "Animal Themes",
    "Animal Wildlife",
    "Swimming",
    "Tentacle",
  ],
  lens: "iPhone 13 mini back dual wide camera 5.1mm f/1.6",
  license: "https://creativecommons.org/licenses/by/4.0/",
  location: "California Academy of Sciences",
  software: "Affinity Photo 2 2.5.7",
  src: "source.jpg",
  state: "California",
  title: "Floating Around",
  variants: [
    {
      description: "Four jellyfish with translucent bodies and long tentacles float gracefully against a vivid background in an aquarium setting (image 1).",
      height: 135,
      src: "post-1.jpg",
      width: 108,
    },
    {
      description: "Four jellyfish with translucent bodies and long tentacles float gracefully against a vivid background in an aquarium setting (image 2).",
      height: 135,
      src: "post-2.jpg",
      width: 108,
    },
    {
      height: 180,
      src: "square.jpg",
      width: 180,
    },
    {
      height: 192,
      src: "story.jpg",
      width: 108,
    },
    {
      height: 180,
      src: "wide.jpg",
      width: 320,
    },
  ],
  width: 240,
}
`;

snapshot[`copyExifToVariants() 1`] = `
{
  camera: "Apple iPhone 13 mini",
  city: "Ankara",
  country: "Turkey",
  date: "2023-02-03T10:27:26",
  description: "A dark-furred cat silhouetted against a window overlooking a snowy landscape.",
  height: 180,
  id: "winter-pause",
  keywords: [
    "Ankara",
    "Turkey",
    "Cat",
    "Window",
    "Silhouette",
    "Dark Fur",
    "Snow",
    "Landscape",
    "Animal",
    "Pet",
    "Domesticated",
    "Feline",
    "House Cat",
    "Winter",
    "Weather",
    "Scenery",
    "View",
    "Outdoor",
    "Nature",
    "Cold",
  ],
  lens: "iPhone 13 mini back dual wide camera 5.1mm f/1.6",
  license: "https://creativecommons.org/licenses/by/4.0/",
  location: "Ankara, Turkey",
  software: "Affinity Photo 2 2.5.7",
  src: "source.jpg",
  state: "Ankara",
  title: "Winter Pause",
  variants: [
    {
      camera: undefined,
      city: undefined,
      country: undefined,
      date: undefined,
      description: undefined,
      height: 144,
      keywords: [],
      lens: undefined,
      license: undefined,
      location: undefined,
      software: undefined,
      src: "post-1.jpg",
      state: undefined,
      title: undefined,
      width: 108,
    },
    {
      camera: undefined,
      city: undefined,
      country: undefined,
      date: undefined,
      description: undefined,
      height: 144,
      keywords: [],
      lens: undefined,
      license: undefined,
      location: undefined,
      software: undefined,
      src: "post-2.jpg",
      state: undefined,
      title: undefined,
      width: 108,
    },
    {
      camera: undefined,
      city: undefined,
      country: undefined,
      date: undefined,
      description: undefined,
      height: 180,
      keywords: [],
      lens: undefined,
      license: undefined,
      location: undefined,
      software: undefined,
      src: "square.jpg",
      state: undefined,
      title: undefined,
      width: 180,
    },
    {
      camera: undefined,
      city: undefined,
      country: undefined,
      date: undefined,
      description: undefined,
      height: 180,
      keywords: [],
      lens: undefined,
      license: undefined,
      location: undefined,
      software: undefined,
      src: "wide.jpg",
      state: undefined,
      title: undefined,
      width: 320,
    },
  ],
  width: 320,
}
`;

snapshot[`copyExifToVariants() 2`] = `
{
  camera: "Apple iPhone 13 mini",
  city: "Ankara",
  country: "Turkey",
  date: "2023-02-03T10:27:26",
  description: "A dark-furred cat silhouetted against a window overlooking a snowy landscape.",
  height: 180,
  id: "winter-pause",
  keywords: [
    "Ankara",
    "Turkey",
    "Cat",
    "Window",
    "Silhouette",
    "Dark Fur",
    "Snow",
    "Landscape",
    "Animal",
    "Pet",
    "Domesticated",
    "Feline",
    "House Cat",
    "Winter",
    "Weather",
    "Scenery",
    "View",
    "Outdoor",
    "Nature",
    "Cold",
  ],
  lens: "iPhone 13 mini back dual wide camera 5.1mm f/1.6",
  license: "https://creativecommons.org/licenses/by/4.0/",
  location: "Ankara, Turkey",
  software: "Affinity Photo 2 2.5.7",
  src: "source.jpg",
  state: "Ankara",
  title: "Winter Pause",
  variants: [
    {
      description: "A dark-furred cat silhouetted against a window overlooking a snowy landscape (image 1).",
      height: 144,
      src: "post-1.jpg",
      width: 108,
    },
    {
      description: "A dark-furred cat silhouetted against a window overlooking a snowy landscape (image 2).",
      height: 144,
      src: "post-2.jpg",
      width: 108,
    },
    {
      height: 180,
      src: "square.jpg",
      width: 180,
    },
    {
      height: 180,
      src: "wide.jpg",
      width: 320,
    },
  ],
  width: 320,
}
`;
